src/ai/backend/common/events/dispatcher.py

```diff
@@ -367,37 +377,69 @@ def unsubscribe(
             cast(EventHandler[Any, AbstractEvent], handler)
         )
 
-    async def handle(self, evh_type: str, evh: EventHandler, source: AgentId, args: tuple) -> None:
+    async def handle(
+        self,
+        evh: EventHandler,
+        source: AgentId,
+        args: tuple,
+        post_callbacks: Sequence[Callable[..., Awaitable[None]]] = tuple(),
+    ) -> None:
         if evh.args_matcher and not evh.args_matcher(args):
             return
         coalescing_opts = evh.coalescing_opts
         coalescing_state = evh.coalescing_state
         cb = evh.callback
+        evh_type = evh.handler_type
         event_cls = evh.event_cls
         if self._closed:
             return
-        if await coalescing_state.rate_control(coalescing_opts):
-            if self._closed:
-                return
-            if self._log_events:
-                log.debug("DISPATCH_{}(evh:{})", evh_type, evh.name)
-            if asyncio.iscoroutinefunction(cb):
-                # mypy cannot catch the meaning of asyncio.iscoroutinefunction().
-                await cb(evh.context, source, event_cls.deserialize(args))  # type: ignore
-            else:
-                cb(evh.context, source, event_cls.deserialize(args))  # type: ignore
+        event_type = event_cls.event_name()
+        start = time.perf_counter()
+        try:
+            if await coalescing_state.rate_control(coalescing_opts):
+                if self._closed:
+                    return
+                if self._log_events:
+                    log.debug("DISPATCH_{}(evh:{})", evh_type.name, evh.name)
+                if asyncio.iscoroutinefunction(cb):
+                    # mypy cannot catch the meaning of asyncio.iscoroutinefunction().
+                    await cb(evh.context, source, event_cls.deserialize(args))  # type: ignore
+                else:
+                    cb(evh.context, source, event_cls.deserialize(args))  # type: ignore
+                for post_callback in post_callbacks:
+                    await post_callback()
+                self._metric_observer.observe_event_success(
+                    event_type=event_type,
+                    duration=time.perf_counter() - start,
+                )
+        except Exception as e:
+            self._metric_observer.observe_event_failure(
+                event_type=event_type,
+                duration=time.perf_counter() - start,
+                exception=e,
+            )
+            log.exception(f"EventDispatcher.{evh_type}(): unexpected-error, {repr(e)}")
+            raise
```

Comment:
Is it okay to raise it here?

Reply:
It is re-raised because of this [test codes](https://github.com/lablup/backend.ai/blob/main/tests/common/test_events.py#L114).
It requires to catch all exceptions raised in the `EventDispatcher.handle()` method.